src/components/bmfont-text.js

```diff
@@ -0,0 +1,333 @@
+      var options = assign({}, data, { font: font, width: textRenderWidth, lineHeight: data.lineHeight || font.common.lineHeight });
+      this.geometry.update(options);
+      this.updateLayout(data);
+    }
+    // if shader changed, update
+    this.updateMaterial(oldData && {shader: oldData.shader, customShader: oldData.customShader});
+  },
+
+  remove: function () {
+    this.geometry.dispose();
+    this.geometry = null;
+    this.el.removeObject3D('bmfont-text');
+  },
+
+  coerceData: function (data) {
+    // We have to coerce some data to numbers/booleans,
+    // as they will be passed directly into text creation and update
+    data = assign({}, data);
+    if (data.lineHeight !== undefined) {
+      data.lineHeight = parseFloat(data.lineHeight);
+      if (!isFinite(data.lineHeight)) { data.lineHeight = undefined; }
+    }
+    if (data.width !== undefined) {
+      data.width = parseFloat(data.width);
+      if (!isFinite(data.width)) { data.width = undefined; }
+    }
+    return data;
+  },
+
+  updateMaterial: function (oldShader) {
+    var data;
+    var changedShader = (oldShader && oldShader.shader) !== this.data.shader || (oldShader && oldShader.customShader) !== this.data.customShader;
+
+    if (changedShader || this.data.customShader) {
+      data = {
+        side: threeSideFromString(this.data.side),
+        transparent: this.data.transparent,
+        alphaTest: this.data.alphaTest,
+        color: this.data.color,
+        opacity: this.data.opacity,
+        map: this.texture
+      };
+    }
+    if (changedShader) {
+      var shader;
+      if (this.data.shader === 'custom') {
+        var ShaderType = shaders[this.data.customShader].Shader;
+        var shaderObject = this.shaderObject = new ShaderType();
+        shaderObject.el = this.el;
+        shaderObject.init(data);
+        shaderObject.update(data);
+        shader = shaderObject.material;
+        shader.transparent = data.transparent; // apparently this is not set on either init or update
+        this.material = shader;
+        if (this.mesh) { this.mesh.material = this.material; }
+        return;
+      }
+      if (this.data.shader === 'SDF') {
+        shader = createSDF(data);
+      } else if (this.data.shader === 'MSDF') {
+        shader = createMSDF(data);
+      } else {
+        shader = createBasic(data);
+      }
+      this.material = new THREE.RawShaderMaterial(shader);
+    } else if (this.data.shader === 'custom') {
+      this.shaderObject.update(data);
+      this.shaderObject.material.transparent = data.transparent; // apparently this is not set on either init or update
+    } else {
+      this.material.uniforms.opacity.value = this.data.opacity;
+      this.material.uniforms.color.value.set(this.data.color);
+      this.material.uniforms.map.value = this.texture;
+    }
+
+    if (this.mesh) { this.mesh.material = this.material; }
+  },
+
+  registerFont: function (key, url) { fontMap[key] = url; },
```

Comment:
registerFont/lookUpFont/loadBMFontPromise/loadTexturePromise methods can also be generic functions